Date picker analytics, restyled analytics

mvkv opened the pull request:
Added date inteval to Analytics endpoint, 
created new css file for analytics page,
Added date picker for selecting analytics interval in frontend,
Restyled analytics page

mvkv committed:
Added date inteval to Analytics endpoint,
created new css file for analytics page,
Added date picker for selecting analytics interval in frontend,
Restyled analytics page

powahftw commented on flask_server.py:
```diff
@@ -232,13 +232,20 @@ def analytics_api_single_user(user_id):
+        user_json_response = []
+        for elem in user_json:
+            if start_date <= elem['taken_at'] <= end_date:
+                user_json_response.append(elem)
```
Use list comprehension or .filter, much cleaner.

mvkv replied:
Done with list comps.

powahftw commented on flask_server.py:
```diff
@@ -232,13 +232,20 @@ def analytics_api_single_user(user_id):
         user_json = json.load(open(json_file_path, 'r'))
-        return jsonify({"selected_user": user_id, "all_users": [], "user_json_file": user_json})
+        start_date = int(user_json[0]['taken_at']) if not start_date else start_date
+        end_date = int(user_json[-1]['taken_at']) if not end_date else end_date
```
If start_date or end_date is not set just keep the story. I would not add a "fake" start_date, end_date that always get picked. It make the logic harder to understand.

You can create a lambda should_keep_story (json)

and put the logic there and then you use filter / list comph. later on with that condition

mvkv replied:
Done

powahftw commented on static/css/analytics.css:
```diff
@@ -0,0 +1,144 @@
+    text-align: center;
+}
+  
+.analitics-users {
```
misspelt analytics ?

mvkv replied:
My bad :D, done

powahftw commented on static/css/analytics.css:
```diff
@@ -0,0 +1,144 @@
+    min-width: 200px;
+    border: solid 1px var(--primary-color-accent);
+    border-radius: 8px;
+
```
nit: empty line

mvkv replied:
Formatted all analytics.css file

powahftw commented on static/css/analytics.css:
```diff
@@ -0,0 +1,144 @@
+    align-items: center;
+    justify-content: center;
+    text-decoration: none;
+    color: white;
```
var(--....) ?

mvkv replied:
Done

powahftw commented on static/css/analytics.css:
```diff
@@ -0,0 +1,144 @@
+    background-color: var(--primary-color-lighter);
+    border: solid 1px var(--primary-color-accent);
+    border-radius: 8px;
+  
```
nit: empty

mvkv replied:
Same as before

powahftw commented on static/css/analytics.css:
```diff
@@ -0,0 +1,144 @@
+.submit-date-btn {
+    margin-top: 8px;
+}
+
```
nit: empty

powahftw commented on templates/analytics.html:
```diff
@@ -1,11 +1,18 @@
-{% set active_page = 'analytics' %}
+{% set active_page = 'analytics' %} 
```
nit: trailing whitespace

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +92,53 @@ const composeStatisticsFromSingleUserJson = (json) => {
+    "Number of videos": nVideos,
+  };
+
+  const mediaChart = `<canvas width="100%" id="chart-media"></canvas>`;
```
The only difference is the id, find a way to only pass the id and generate the rest (<canvas width="100%" id="XXXX"></canvas>)

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +92,53 @@ const composeStatisticsFromSingleUserJson = (json) => {
+
+  const mediaChart = `<canvas width="100%" id="chart-media"></canvas>`;
+  const freqChart = `<canvas width="100%" id="chart-hourly-freq"></canvas>`;
+  const charts = {
```
Chart is used only at line 118, move it closer to where it get used so it's clearer

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +92,53 @@ const composeStatisticsFromSingleUserJson = (json) => {
+
+  let userDetailsBar = "";
+  userDetailsBar += `<div class="user-details-bar">`;
+  for (const [key, value] of Object.entries(userDetails)) {
```
Ofc they are [key, value], but use better naming for this context.
Same below

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -28,13 +28,29 @@ const composeUserChoicePage = (users) => {
   return `<div class='analitics-wrapper'>${pageHtml}</div>`;
 };
 
+const composeDatePicker = () =>
```
Just add this to the HTML directly instead of generating it on the JS side, it doesn't change. Make it default hidden and you can still use the condition at line 291 to make it visible if you want to have this only on successful responses.

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -215,11 +282,11 @@ const renderChartsFromSingleUserJson = (json) => {
-const getAndRenderAnalytics = async () => {
-
-  const requestUrl = `${baseUrl}/${API_PREFIX}${pathName}`;
+const getAndRenderAnalytics = async (startDate, endDate) => {
```
Make startDate and endDate an optional  param, this simplify the logic also below(eg line 335 / 325)

mvkv replied:
I think this is related with the misunderstanding about the "|" operator

powahftw replied:
Nope, different thing. I'd make it so you are not required to pass both param when calling it.

mvkv replied:
Made optional params, also added default values and rendering of default values in the date picker.

powahftw commented on static/js/analytics.js:
```diff
@@ -239,10 +306,32 @@ const getAndRenderAnalytics = async () => {
+      const endDate = new Date(document.getElementById("end-date").value);
+      endDate.setDate(endDate.getDate() + 1);
+      const startDateTimestamp = (startDate.getTime() / 1000) | 0;
+      const endDateTimestamp = (endDate.getTime() / 1000) | 0;
```
0 is not the same as null / undefined, check if that's what you really want.

mvkv replied:
The single "|" is used to trim the decimal part, it's not the or operator

powahftw replied:
Ah gotcha. I'd use something more commonly known, such as Math.trunc()

mvkv replied:
Done

mvkv committed:
Moved date-picker rendering to html,
creating canvas through a function in analytics,
improved some variable names,
using list comph instead for loops in flask_server,

mvkv committed:
Removed default values for date picker in backend

powahftw commented on static/css/style.css:
```diff
@@ -45,6 +45,7 @@
   --red-complementary-color: #f72823;
   --blue-complementary-color: #4be0fe;
   --green-complementary-color: #23f751;
+  --white: #FFFFFF;
```
nit: We used lowercase for all other color, I'd follow the same convention for this as well.

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +76,51 @@ const composeStatisticsFromSingleUserJson = (json) => {
 
-  if (NGPSLocations > 0) {
-    pageHtml += `<p>Geotagged ${NGPSLocations} times</p>`;
+  const userDetails = {
```
I'd rename userDetails to something clearer, It's not details about an user but about the current processed results.

eg: userDetails -> userSelectedStoriesInfo

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +76,51 @@ const composeStatisticsFromSingleUserJson = (json) => {
+  }
+  userDetailsBar += `</div>`;
+
+  const chartCanvas = (id) => `<canvas width="100%" id="${id}"></canvas>`;
```
createCanvasElFromId instead of chartCanvas?

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +76,51 @@ const composeStatisticsFromSingleUserJson = (json) => {
+    "Media chart": chartCanvas("chart-media"),
+    "Frequency chart": chartCanvas("chart-hourly-freq"),
+  };
+  let analyticsBox = "";
```
no reason to have it empty at the start if you have already stuff on the line below. 
Do
`let analyticsBox += `<div class="analytics-box-container">`;`

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -64,24 +76,51 @@ const composeStatisticsFromSingleUserJson = (json) => {
+    "Number of videos": nVideos,
+  };
+
+  let userDetailsBar = "";
```
Same as suggested below. (btw pageHtml above is slightly different so i'd leave that unchanged)

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -236,13 +285,35 @@ const getAndRenderAnalytics = async () => {
 
-window.onload = () => {
-  getAndRenderAnalytics();
+const dateFilter = () => {
```
Use a more descriptive name, something like dateFilter -> addDateFilterEventListener

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -236,13 +285,35 @@ const getAndRenderAnalytics = async () => {
+
+window.onload = async () => {
+  const currentDate = new Date();
+  const endDate = (currentDate.getTime() / 1000) | 0;
```
Remove these 2 line after making the arguments optional

mvkv replied:
Moved them into getAndRenderAnalytics in order to handle default interval of 3 months

powahftw commented on templates/analytics.html:
```diff
@@ -1,11 +1,33 @@
+    </div>
+    <button class="button submit-date-btn" id="submit-date-btn">Filter</button>
+  </div>
+
```
nit: empty line

mvkv replied:
Formatted the whole document with Prettier

mvkv committed:
Added default values in date picker,
now date params are optional in getAndRenderAnalyics,
formatted analytics.html with Prettier,

mvkv committed:
Added func to retrieve date interval and render it

powahftw commented on static/js/analytics.js:
```diff
@@ -265,26 +265,35 @@ const renderChartsFromSingleUserJson = (json) => {
+  let endDate = endDateField ? new Date(endDateField) : new Date();
+  let startDate = startDateField
+    ? new Date(startDateField)
+    : new Date(new Date().setMonth(endDate.getMonth() - monthsAgo));
```
I think you don't need the outer new Date() here

mvkv replied:
It was needed, but now I changed the logic, so it's not useful anymore

powahftw commented on static/js/analytics.js:
```diff
@@ -265,26 +265,35 @@ const renderChartsFromSingleUserJson = (json) => {
+const getDateIntervalAndUpdateDatePicker = () => {
+  const startDateField = document.getElementById("start-date").value;
+  const endDateField = document.getElementById("end-date").value;
+  const monthsAgo = 3;
```
FULL_CAPS as it's a constant, also I'd move it to the start. and call it differently.

eg: DEFAULT_LOOKBACK_MONTHS

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -265,26 +265,35 @@ const renderChartsFromSingleUserJson = (json) => {
+    ? new Date(startDateField)
+    : new Date(new Date().setMonth(endDate.getMonth() - monthsAgo));
+
+  endDate.setDate(endDate.getDate() + 1); // Handles the last day
```
If every time this get called you update the last day with +1 you can keep going forward in time. Instead move this part of the logic to the comparison function on the Python API side.

mvkv replied:
Moved logic on backend

powahftw commented on static/js/analytics.js:
```diff
@@ -265,26 +265,35 @@ const renderChartsFromSingleUserJson = (json) => {
+    : new Date(new Date().setMonth(endDate.getMonth() - monthsAgo));
+
+  endDate.setDate(endDate.getDate() + 1); // Handles the last day
+  startDate = startDate.toISOString().split("T")[0];
+  endDate = endDate.toISOString().split("T")[0];
+  const startDateTimestamp = Date.parse(startDate) / 1000;
+  const endDateTimestamp = Date.parse(endDate) / 1000;
```
Move this part of logic to updateDatePicker, so it accept 2 Date params and handle inside how to convert it to the YYYY-MM-DD format

mvkv replied:
Done

powahftw commented on static/js/analytics.js:
```diff
@@ -265,26 +265,35 @@ const renderChartsFromSingleUserJson = (json) => {
+  const startDateTimestamp = Date.parse(startDate) / 1000;
+  const endDateTimestamp = Date.parse(endDate) / 1000;
+
+  updateDatePicker(startDate, endDate);
```
Call updateDatePicker only if one / both of startDateField / endDateField is unset. Otherwise there is no need to set anything as it's already good!

mvkv replied:
Done

mvkv committed:
Moved conversion from timestamp to date to updateDatePicker function,
moved logic to handle last day backend sid\e,
moved normalizing timestamp backend side.

powahftw commented on static/js/analytics.js:
```diff
@@ -265,27 +267,26 @@ const renderChartsFromSingleUserJson = (json) => {
-
-  updateDatePicker(startDate, endDate);
+  const currentDate = new Date();
+  let endDateTimestamp = endDateField
```
I'd avoid using the same name of the arg.

powahftw replied:
Also this could be a const

mvkv replied:
Done

mvkv committed:
Changed timestamps to consts
